src/libxtp/gwbse/bse_operator.cc

```diff
@@ -58,48 +58,59 @@ Eigen::MatrixXd BSE_OPERATOR<cqp, cx, cd, cd2>::matmul(
                                 auxsize);
   }
 
-#pragma omp parallel for schedule(dynamic)
-  for (Index c1 = 0; c1 < _bse_ctotal; c1++) {
+#pragma omp parallel
+  {
+    Index threadid = OPENMP::getThreadId();
+#pragma omp for schedule(dynamic)
+    for (Index c1 = 0; c1 < _bse_ctotal; c1++) {
 
-    Eigen::MatrixXd Temp;
-    if (cd != 0) {
-      Temp = -cd * (_Mmn[c1 + cmin].block(cmin, 0, _bse_ctotal, auxsize));
-      transform.PrepareMatrix1(Temp);
-    }
-    if (cd2 != 0) {
-      Temp = -cd2 * (_Mmn[c1 + cmin].block(vmin, 0, _bse_vtotal, auxsize));
-      transform.PrepareMatrix1(Temp);
-    }
-
-    for (Index v1 = 0; v1 < _bse_vtotal; v1++) {
-      transform.SetTempZero();
+      Eigen::MatrixXd Temp;
       if (cd != 0) {
-        transform.PrepareMatrix2(
-            _Mmn[v1 + vmin].block(vmin, 0, _bse_vtotal, auxsize), cd2 != 0);
+        Temp = -cd * (_Mmn[c1 + cmin].block(cmin, 0, _bse_ctotal, auxsize));
+        transform.PrepareMatrix1(Temp, threadid);
       }
       if (cd2 != 0) {
-        transform.PrepareMatrix2(
-            _Mmn[v1 + vmin].block(cmin, 0, _bse_ctotal, auxsize), cd2 != 0);
+        Temp = -cd2 * (_Mmn[c1 + cmin].block(vmin, 0, _bse_vtotal, auxsize));
+        transform.PrepareMatrix1(Temp, threadid);
       }
```

Comment:
I don't really understand the logic here, if both cd and cd2 are nonzero then only the last statement has any effect and the first operation could have been left out . An if-else or if-elif-else statement would seem more appropriate in that case, the same applies to a few lines below where a similar control flow is used.

Reply:
there is a `static_assert` a little higher which checks that cd and cd2 can never both be nonzero, because that makes no sense from the physics side.

Reply:
I could use a ìf`  `else if` statement, but I wanted to keep it a bit more symmetric, but I see your point.

Reply:
otherwise I agree I could have used an `if` `else if` case, but I thought the more symmetric thing makes more sense. I can change that, if you think it makes the code more intuitive.

Reply:
Okay, I missed the static assert, that at least makes it logical, but then even an if-else would be enough and it would immediately signal that one case excludes the other. 

Change it if you like, it is a minor detail and mostly a taste thing, so if you leave it, it is also fine.